src/control/LatexController.cpp

```diff
@@ -110,147 +129,202 @@ void LatexController::findSelectedTexElement()
-	layer = page->getSelectedLayer();
+	this->page = this->doc->getPage(pageNr);
+	this->layer = page->getSelectedLayer();
 
-	selectedTexImage = view->getSelectedTex();
-	selectedText = view->getSelectedText();
+	this->selectedTexImage = view->getSelectedTex();
+	this->selectedText = view->getSelectedText();
 
-	if (selectedTexImage || selectedText)
+	if (this->selectedTexImage || this->selectedText)
 	{
 		// this will get the position of the Latex properly
 		EditSelection *theSelection = control->getWindow()->getXournal()->getSelection();
 		posx = theSelection->getXOnView();
 		posy = theSelection->getYOnView();
 
-		if (selectedTexImage)
+		if (this->selectedTexImage != nullptr)
 		{
-			initalTex = selectedTexImage->getText();
-			imgwidth = selectedTexImage->getElementWidth();
-			imgheight = selectedTexImage->getElementHeight();
+			initalTex = this->selectedTexImage->getText();
+			imgwidth = this->selectedTexImage->getElementWidth();
+			imgheight = this->selectedTexImage->getElementHeight();
 		}
 		else
 		{
 			initalTex += "\\text{";
-			initalTex += selectedText->getText();
+			initalTex += this->selectedText->getText();
 			initalTex += "}";
-			imgwidth = selectedText->getElementWidth();
-			imgheight = selectedText->getElementHeight();
+			imgwidth = this->selectedText->getElementWidth();
+			imgheight = this->selectedText->getElementHeight();
 		}
 	}
-	if (initalTex.empty())
+	if (this->initalTex.empty())
 	{
-		initalTex = "x^2";
+		this->initalTex = "x^2";
 	}
-	currentTex = initalTex;
-	doc->unlock();
+	this->currentTex = this->initalTex;
+	this->doc->unlock();
 
 	// need to do this otherwise we can't remove the image for its replacement
-	control->clearSelectionEndText();
+	this->control->clearSelectionEndText();
 }
 
 void LatexController::showTexEditDialog()
 {
 	XOJ_CHECK_TYPE(LatexController);
 
-	dlg = new LatexDialog(control->getGladeSearchPath());
+	this->dlg = new LatexDialog(control->getGladeSearchPath());
 
 	// For 'real time' LaTex rendering in the dialog
-	dlg->setTex(initalTex);
+	this->dlg->setTex(initalTex);
 	g_signal_connect(dlg->getTextBuffer(), "changed", G_CALLBACK(handleTexChanged), this);
-	
-	
-	// The controller owns the tempRender because, on signal changed, he has to handle the old/new renders
-	if (temporaryRender != NULL)
+
+
+	// The controller owns the tempRender because, on signal changed, it has to handle the old/new renders
+	if (this->temporaryRender != nullptr)
 	{
-		dlg->setTempRender(temporaryRender->getPdf(), initalTex.size());
+		this->dlg->setTempRender(this->temporaryRender->getPdf(), initalTex.size());
 	}
 
-	dlg->show(GTK_WINDOW(control->getWindow()->getWindow()));
+	this->dlg->show(GTK_WINDOW(control->getWindow()->getWindow()));
 
-	deletePreviousRender();
-	currentTex = dlg->getTex();
+	currentTex = this->dlg->getTex();
 	currentTex += " ";
 
-	delete dlg;
+	delete this->dlg;
+}
+
+void LatexController::triggerImageUpdate()
+{
+	if (this->isUpdating)
+	{
+		return;
+	}
```

Comment:
It would be better to abort the current image generation and create a new image with the newer data. That could fix the issue with the generating lag. I also see a problem when you type the last character while the image is currently updating.

Reply:
On the first point: this doesn't offer any significant advantage to the user experience. This means that if is the user is rapidly typing in the latex, then the user won't be able to see the preview until he or she is done typing. Also, aborting is a bit hard here with the possible race conditions we can have. From the few hours I've been using the new implementation, the lag hasn't been too noticeable.

On the second point: in the callback, after rendering the preview, we check to see if the latex has been updated since the last time an update was triggered. If it has, we trigger another update (see bottom of `onPdfRenderComplete`).

Reply:
Ahh okay then you thought about these cases. All good for me!